src/Nethermind/Nethermind.Evm/EvmObjectFormat/EofCodeValidator.cs

```diff
@@ -0,0 +1,903 @@
+    struct Worklet
+    {
+        public Worklet(ushort position, ushort stackHeight)
+        {
+            Position = position;
+            StackHeight = stackHeight;
+        }
+        public ushort Position;
+        public ushort StackHeight;
+    }
+
+    private interface IEofVersionHandler
+    {
+        bool ValidateBody(ReadOnlySpan<byte> code, EofHeader header);
+        bool TryParseEofHeader(ReadOnlySpan<byte> code, [NotNullWhen(true)] out EofHeader? header);
+    }
+
+    // magic prefix : EofFormatByte is the first byte, EofFormatDiff is chosen to diff from previously rejected contract according to EIP3541
+    public static byte[] MAGIC = { 0xEF, 0x00 };
+    public const byte ONE_BYTE_LENGTH = 1;
+    public const byte TWO_BYTE_LENGTH = 2;
+    public const byte VERSION_OFFSET = TWO_BYTE_LENGTH; // magic lenght
+
+    private static readonly Dictionary<byte, IEofVersionHandler> _eofVersionHandlers = new();
+    internal static ILogger Logger { get; set; } = NullLogger.Instance;
+
+    static EvmObjectFormat()
+    {
+        _eofVersionHandlers.Add(Eof1.VERSION, new Eof1());
+    }
+
+    /// <summary>
+    /// returns whether the code passed is supposed to be treated as Eof regardless of its validity.
+    /// </summary>
+    /// <param name="container">Machine code to be checked</param>
+    /// <returns></returns>
+    public static bool IsEof(ReadOnlySpan<byte> container, [NotNullWhen(true)] out int version)
+    {
+        if(container.Length >= MAGIC.Length + 1)
+        {
+            version = container[MAGIC.Length];
+            return container.StartsWith(MAGIC);
+        } else
+        {
+            version = 0;
+            return false;
+        }
+        
+    }
+
+    public static bool IsEofn(ReadOnlySpan<byte> container, byte version) => container.Length >= MAGIC.Length + 1 && container.StartsWith(MAGIC) && container[MAGIC.Length] == version;
+
+    public static bool IsValidEof(ReadOnlySpan<byte> container, bool validateSubContainers, [NotNullWhen(true)] out EofHeader? header)
+    {
+        if (container.Length > VERSION_OFFSET
+            && _eofVersionHandlers.TryGetValue(container[VERSION_OFFSET], out IEofVersionHandler handler)
+            && handler.TryParseEofHeader(container, out header))
+        {
+            EofHeader h = header.Value;
+            if (handler.ValidateBody(container, h))
+            {
+                return true;
+            }
+
+            if(validateSubContainers && header?.ContainerSection?.Count > 0)
+            {
+                int containerSize = header.Value.ContainerSection.Value.Count;
+                byte[][] containers = ArrayPool<byte[]>.Shared.Rent(containerSize);
+
+                for (int i = 0; i < containerSize; i++)
+                {
+                    containers[i] = container.Slice(header.Value.ContainerSection.Value.Start + header.Value.ContainerSection.Value[i].Start, header.Value.ContainerSection.Value[i].Size).ToArray();
```

Comment:
Why materialize them? Wouldn't remembering (offset, length) work just fine? This could be `ArrayPool<(short, short)>` or something like this and would be later sliceable to a span. Again, not sure how how it is.

Also, can't we validate container right after it's created so that the loops, for and foreach become one to have it single pass?

Reply:
yup, you're right, or even better using the header to slice the subcontainer since it already has the offsets and sizes